File: SlidingWindow/Java/SlidingWindowSolution.java
package SlidingWindow.Java;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.Deque;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;

/**
 * Complete implementation of Sliding Window algorithms.
 * This class provides a reference solution for the SlidingWindow blueprint.
 * You can use this as a guide if you get stuck with your implementation.
 */
public class SlidingWindowSolution {
    
    /**
     * Find the maximum sum subarray of size k.
     * 
     * @param arr The input array
     * @param k The size of the window
     * @return The maximum sum of any subarray of size k
     */
    public int findMaxSumSubarray(int[] arr, int k) {
        if (arr == null || arr.length == 0 || k <= 0 || k > arr.length) {
            return 0;
        }
        
        int maxSum = 0;
        int windowSum = 0;
        
        // Calculate sum of first k elements
        for (int i = 0; i < k; i++) {
            windowSum += arr[i];
        }
        
        maxSum = windowSum;
        
        // Slide the window and update maxSum
        for (int i = k; i < arr.length; i++) {
            windowSum = windowSum - arr[i - k] + arr[i];
            maxSum = Math.max(maxSum, windowSum);
        }
        
        return maxSum;
    }
    
    /**
     * Find the smallest subarray with a sum greater than or equal to the target.
     * 
     * @param arr The input array
     * @param target The target sum
     * @return The length of the smallest subarray with sum >= target, or 0 if no such subarray exists
     */
    public int findMinSubarrayLength(int[] arr, int target) {
        if (arr == null || arr.length == 0 || target <= 0) {
            return 0;
        }
        
        int minLength = Integer.MAX_VALUE;
        int windowSum = 0;
        int windowStart = 0;
        
        for (int windowEnd = 0; windowEnd < arr.length; windowEnd++) {
            windowSum += arr[windowEnd];
            
            // Shrink the window as small as possible while maintaining the sum >= target
            while (windowSum >= target) {
                minLength = Math.min(minLength, windowEnd - windowStart + 1);
                windowSum -= arr[windowStart];
                windowStart++;
            }
        }
        
        return minLength == Integer.MAX_VALUE ? 0 : minLength;
    }
    
    /**
     * Find the longest substring with at most k distinct characters.
     * 
     * @param str The input string
     * @param k The maximum number of distinct characters
     * @return The length of the longest substring with at most k distinct characters
     */
    public int findLongestSubstringKDistinct(String str, int k) {
        if (str == null || str.length() == 0 || k <= 0) {
            return 0;
        }
        
        int maxLength = 0;
        int windowStart = 0;
        Map<Character, Integer> charFrequency = new HashMap<>();
        
        for (int windowEnd = 0; windowEnd < str.length(); windowEnd++) {
            char rightChar = str.charAt(windowEnd);
            charFrequency.put(rightChar, charFrequency.getOrDefault(rightChar, 0) + 1);
            
            // Shrink the window if we have more than k distinct characters
            while (charFrequency.size() > k) {
                char leftChar = str.charAt(windowStart);
                charFrequency.put(leftChar, charFrequency.get(leftChar) - 1);
                if (charFrequency.get(leftChar) == 0) {
                    charFrequency.remove(leftChar);
                }
                windowStart++;
            }
            
            maxLength = Math.max(maxLength, windowEnd - windowStart + 1);
        }
        
        return maxLength;
    }
    
    /**
     * Find the longest substring without repeating characters.
     * 
     * @param str The input string
     * @return The length of the longest substring without repeating characters
     */
    public int findLongestSubstringNoRepeat(String str) {
        if (str == null || str.length() == 0) {
            return 0;
        }
        
        int maxLength = 0;
        int windowStart = 0;
        Map<Character, Integer> charIndexMap = new HashMap<>();
        
        for (int windowEnd = 0; windowEnd < str.length(); windowEnd++) {
            char rightChar = str.charAt(windowEnd);
            
            // If the character is already in the window, shrink the window
            if (charIndexMap.containsKey(rightChar)) {
                // Move the window start to the right of the last occurrence of the character
                windowStart = Math.max(windowStart, charIndexMap.get(rightChar) + 1);
            }
            
            // Update the index of the character
            charIndexMap.put(rightChar, windowEnd);
            
            // Update the maximum length
            maxLength = Math.max(maxLength, windowEnd - windowStart + 1);
        }
        
        return maxLength;
    }
    
    /**
     * Find all anagrams of a pattern in a string.
     * 
     * @param str The input string
     * @param pattern The pattern to find anagrams of
     * @return An array of starting indices of all anagrams of the pattern in the string
     */
    public int[] findAnagrams(String str, String pattern) {
        if (str == null || pattern == null || str.length() < pattern.length()) {
            return new int[0];
        }
        
        List<Integer> resultIndices = new ArrayList<>();
        Map<Character, Integer> patternFrequency = new HashMap<>();
        Map<Character, Integer> windowFrequency = new HashMap<>();
        
        // Build the frequency map for the pattern
        for (char c : pattern.toCharArray()) {
            patternFrequency.put(c, patternFrequency.getOrDefault(c, 0) + 1);
        }
        
        int windowStart = 0;
        int matched = 0;
        
        for (int windowEnd = 0; windowEnd < str.length(); windowEnd++) {
            char rightChar = str.charAt(windowEnd);
            
            // Add the right character to the window frequency map
            windowFrequency.put(rightChar, windowFrequency.getOrDefault(rightChar, 0) + 1);
            
            // If the frequency of the current character matches with the pattern, increment the matched count
            if (patternFrequency.containsKey(rightChar) && windowFrequency.get(rightChar).equals(patternFrequency.get(rightChar))) {
                matched++;
            }
            
            // If we have found an anagram
            if (matched == patternFrequency.size()) {
                resultIndices.add(windowStart);
            }
            
            // Shrink the window if its size is equal to the pattern length
            if (windowEnd >= pattern.length() - 1) {
                char leftChar = str.charAt(windowStart);
                
                // If the frequency of the left character matches with the pattern, decrement the matched count
                if (patternFrequency.containsKey(leftChar) && windowFrequency.get(leftChar).equals(patternFrequency.get(leftChar))) {
                    matched--;
                }
                
                // Remove the left character from the window frequency map
                windowFrequency.put(leftChar, windowFrequency.get(leftChar) - 1);
                if (windowFrequency.get(leftChar) == 0) {
                    windowFrequency.remove(leftChar);
                }
                
                windowStart++;
            }
        }
        
        // Convert the list to an array
        int[] result = new int[resultIndices.size()];
        for (int i = 0; i < resultIndices.size(); i++) {
            result[i] = resultIndices.get(i);
        }
        
        return result;
    }
    
    /**
     * Find the maximum of each subarray of size k.
     * 
     * @param arr The input array
     * @param k The size of the window
     * @return An array containing the maximum of each subarray of size k
     */
    public int[] findMaxInSlidingWindow(int[] arr, int k) {
        if (arr == null || arr.length == 0 || k <= 0 || k > arr.length) {
            return new int[0];
        }
        
        int[] result = new int[arr.length - k + 1];
        Deque<Integer> deque = new ArrayDeque<>();
        
        for (int i = 0; i < arr.length; i++) {
            // Remove elements outside the current window
            while (!deque.isEmpty() && deque.peekFirst() < i - k + 1) {
                deque.pollFirst();
            }
            
            // Remove elements smaller than the current element
            while (!deque.isEmpty() && arr[deque.peekLast()] < arr[i]) {
                deque.pollLast();
            }
            
            // Add the current element
            deque.offerLast(i);
            
            // Add the maximum element to the result
            if (i >= k - 1) {
                result[i - k + 1] = arr[deque.peekFirst()];
            }
        }
        
        return result;
    }
    
    /**
     * Find the minimum window substring that contains all characters of the pattern.
     * 
     * @param str The input string
     * @param pattern The pattern string
     * @return The minimum window substring, or an empty string if no such substring exists
     */
    public String findMinWindowSubstring(String str, String pattern) {
        if (str == null || pattern == null || str.length() < pattern.length()) {
            return "";
        }
        
        Map<Character, Integer> patternFrequency = new HashMap<>();
        Map<Character, Integer> windowFrequency = new HashMap<>();
        
        // Build the frequency map for the pattern
        for (char c : pattern.toCharArray()) {
            patternFrequency.put(c, patternFrequency.getOrDefault(c, 0) + 1);
        }
        
        int windowStart = 0;
        int matched = 0;
        int minLength = Integer.MAX_VALUE;
        int substrStart = 0;
        
        for (int windowEnd = 0; windowEnd < str.length(); windowEnd++) {
            char rightChar = str.charAt(windowEnd);
            
            // Add the right character to the window frequency map
            windowFrequency.put(rightChar, windowFrequency.getOrDefault(rightChar, 0) + 1);
            
            // If the frequency of the current character matches with the pattern, increment the matched count
            if (patternFrequency.containsKey(rightChar) && windowFrequency.get(rightChar) <= patternFrequency.get(rightChar)) {
                matched++;
            }
            
            // Shrink the window if we have found all characters
            while (matched == pattern.length()) {
                // Update the minimum window
                if (windowEnd - windowStart + 1 < minLength) {
                    minLength = windowEnd - windowStart + 1;
                    substrStart = windowStart;
                }
                
                char leftChar = str.charAt(windowStart);
                
                // Remove the left character from the window frequency map
                windowFrequency.put(leftChar, windowFrequency.get(leftChar) - 1);
                
                // If the frequency of the left character is less than the pattern, decrement the matched count
                if (patternFrequency.containsKey(leftChar) && windowFrequency.get(leftChar) < patternFrequency.get(leftChar)) {
                    matched--;
                }
                
                windowStart++;
            }
        }
        
        return minLength == Integer.MAX_VALUE ? "" : str.substring(substrStart, substrStart + minLength);
    }
    
    /**
     * Find the maximum profit by buying and selling a stock once.
     * 
     * @param prices The array of stock prices
     * @return The maximum profit
     */
    public int findMaxProfit(int[] prices) {
        if (prices == null || prices.length < 2) {
            return 0;
        }
        
        int maxProfit = 0;
        int minPrice = prices[0];
        
        for (int i = 1; i < prices.length; i++) {
            // Update the minimum price
            minPrice = Math.min(minPrice, prices[i]);
            
            // Update the maximum profit
            maxProfit = Math.max(maxProfit, prices[i] - minPrice);
        }
        
        return maxProfit;
    }
    
    /**
     * Find the longest subarray with ones after replacing at most k zeros.
     * 
     * @param arr The input array (containing only 0s and 1s)
     * @param k The maximum number of zeros that can be replaced
     * @return The length of the longest subarray with ones after replacing at most k zeros
     */
    public int findLongestOnesSubarray(int[] arr, int k) {
        if (arr == null || arr.length == 0 || k < 0) {
            return 0;
        }
        
        int maxLength = 0;
        int windowStart = 0;
        int zeroCount = 0;
        
        for (int windowEnd = 0; windowEnd < arr.length; windowEnd++) {
            // If the current element is 0, increment the zero count
            if (arr[windowEnd] == 0) {
                zeroCount++;
            }
            
            // Shrink the window if we have more than k zeros
            while (zeroCount > k) {
                if (arr[windowStart] == 0) {
                    zeroCount--;
                }
                windowStart++;
            }
            
            // Update the maximum length
            maxLength = Math.max(maxLength, windowEnd - windowStart + 1);
        }
        
        return maxLength;
    }
}

File: SlidingWindow/Python/sliding_window_solution.py
"""
Complete implementation of Sliding Window algorithms.
This class provides a reference solution for the Sliding Window blueprint.
You can use this as a guide if you get stuck with your implementation.
"""
from collections import defaultdict, deque

class SlidingWindowSolution:
    def __init__(self):
        """
        Constructor to create a new SlidingWindow instance.
        """
        pass
    
    def max_sum_subarray(self, arr, k):
        """
        Find the maximum sum of a subarray of size k.
        
        Args:
            arr: Input array of integers
            k: Size of the sliding window
            
        Returns:
            Maximum sum of a subarray of size k
        """
        if not arr or k <= 0 or k > len(arr):
            return 0
        
        # Initialize variables
        max_sum = 0
        window_sum = 0
        
        # Calculate sum of first window
        for i in range(k):
            window_sum += arr[i]
        
        max_sum = window_sum
        
        # Slide the window and update max_sum
        for i in range(k, len(arr)):
            window_sum = window_sum + arr[i] - arr[i - k]
            max_sum = max(max_sum, window_sum)
        
        return max_sum
    
    def min_size_subarray_sum(self, arr, target_sum):
        """
        Find the minimum size subarray with a sum at least target_sum.
        
        Args:
            arr: Input array of positive integers
            target_sum: Target sum
            
        Returns:
            Minimum size of a subarray with sum at least target_sum, or 0 if no such subarray exists
        """
        if not arr or target_sum <= 0:
            return 0
        
        # Initialize variables
        min_length = float('inf')
        window_sum = 0
        start = 0
        
        # Slide the window
        for end in range(len(arr)):
            window_sum += arr[end]
            
            # Shrink the window as small as possible while maintaining the sum >= target_sum
            while window_sum >= target_sum:
                min_length = min(min_length, end - start + 1)
                window_sum -= arr[start]
                start += 1
        
        return min_length if min_length != float('inf') else 0
    
    def longest_substring_without_repeating_chars(self, s):
        """
        Find the length of the longest substring without repeating characters.
        
        Args:
            s: Input string
            
        Returns:
            Length of the longest substring without repeating characters
        """
        if not s:
            return 0
        
        # Initialize variables
        max_length = 0
        start = 0
        char_index_map = {}
        
        # Slide the window
        for end in range(len(s)):
            # If the character is already in the window, update the start pointer
            if s[end] in char_index_map and char_index_map[s[end]] >= start:
                start = char_index_map[s[end]] + 1
            
            # Update the character's index
            char_index_map[s[end]] = end
            
            # Update max_length
            max_length = max(max_length, end - start + 1)
        
        return max_length
    
    def find_all_anagrams(self, s, p):
        """
        Find all start indices of anagrams of p in s.
        
        Args:
            s: Input string
            p: Pattern string
            
        Returns:
            List of start indices of anagrams of p in s
        """
        if not s or not p or len(s) < len(p):
            return []
        
        # Initialize variables
        result = []
        p_count = defaultdict(int)
        window_count = defaultdict(int)
        
        # Count characters in pattern
        for char in p:
            p_count[char] += 1
        
        # Initialize window
        for i in range(len(p)):
            window_count[s[i]] += 1
        
        # Check if the initial window is an anagram
        if window_count == p_count:
            result.append(0)
        
        # Slide the window
        for i in range(len(p), len(s)):
            # Add the new character to the window
            window_count[s[i]] += 1
            
            # Remove the leftmost character from the window
            window_count[s[i - len(p)]] -= 1
            
            # If the count becomes 0, remove the character from the dictionary
            if window_count[s[i - len(p)]] == 0:
                del window_count[s[i - len(p)]]
            
            # Check if the current window is an anagram
            if window_count == p_count:
                result.append(i - len(p) + 1)
        
        return result
    
    def max_sliding_window(self, nums, k):
        """
        Find the maximum element in each sliding window of size k.
        
        Args:
            nums: Input array of integers
            k: Size of the sliding window
            
        Returns:
            Array of maximum elements in each sliding window
        """
        if not nums or k <= 0 or k > len(nums):
            return []
        
        # Initialize variables
        result = []
        deq = deque()  # Store indices of elements in decreasing order
        
        for i in range(len(nums)):
            # Remove elements outside the window
            while deq and deq[0] < i - k + 1:
                deq.popleft()
            
            # Remove smaller elements as they are not useful
            while deq and nums[deq[-1]] < nums[i]:
                deq.pop()
            
            # Add current element
            deq.append(i)
            
            # Add maximum element to result if we have a complete window
            if i >= k - 1:
                result.append(nums[deq[0]])
        
        return result
    
    def longest_repeating_character_replacement(self, s, k):
        """
        Find the length of the longest substring containing the same letter after replacing at most k characters.
        
        Args:
            s: Input string
            k: Maximum number of characters to replace
            
        Returns:
            Length of the longest substring containing the same letter after replacing at most k characters
        """
        if not s:
            return 0
        
        # Initialize variables
        max_length = 0
        max_count = 0
        char_count = defaultdict(int)
        start = 0
        
        # Slide the window
        for end in range(len(s)):
            # Update character count
            char_count[s[end]] += 1
            
            # Update max_count (most frequent character in the window)
            max_count = max(max_count, char_count[s[end]])
            
            # If the number of characters to replace exceeds k, shrink the window
            if end - start + 1 - max_count > k:
                char_count[s[start]] -= 1
                start += 1
            
            # Update max_length
            max_length = max(max_length, end - start + 1)
        
        return max_length
